packages/cli-lib/lib/processFieldsJs.js

```diff
@@ -0,0 +1,123 @@
+const fs = require('fs');
+const semver = require('semver');
+const { pathToFileURL } = require('url');
+const { getExt } = require('../path');
+const { logger, setLogLevel } = require('../logger');
+const { FieldErrors, logFieldsJsError } = require('../errorHandlers');
+const { i18n } = require('./lang');
+
+const i18nKey = 'cli.commands.upload';
+
+const { dirName, fieldOptions, filePath, writeDir, logLevel } = process.env;
+const baseName = path.basename(filePath);
+
+// We need to carry over the log level from the parent instance
+setLogLevel(Number(logLevel));
+
+const logError = (err, info = {}) => logFieldsJsError(err, filePath, info);
+const errorCatch = e => {
+  logError(e);
+  // Errors caught by this could be caused by the users javascript, so just print the whole error for them.
+  logger.error(e);
+};
+
+logger.info(
+  i18n(`${i18nKey}.converting`, {
+    src: dirName + `/${baseName}`,
+    dest: dirName + '/fields.json',
+  })
+);
+
+/*
+ * How this works: dynamicImport() will always return either a Promise or undefined.
+ * In the case when it's a Promise, its expected that it will resolve to a function.
+ * This function has optional return type of Promise<Array> | Array. In order to have uniform handling,
+ * we wrap the return value of the function in a Promise.resolve(), and then process.
+ */
+
+const fieldsPromise = dynamicImport(filePath).catch(e => errorCatch(e));
+
+fieldsPromise.then(fieldsFunc => {
+  const fieldsFuncType = typeof fieldsFunc;
+  if (fieldsFuncType !== 'function') {
+    logError(FieldErrors.IsNotFunction, {
+      returned: fieldsFuncType,
+    });
+    return;
+  }
+  return Promise.resolve(fieldsFunc(fieldOptions)).then(fields => {
+    if (!Array.isArray(fields)) {
+      logError(FieldErrors.DoesNotReturnArray, {
+        returned: typeof fields,
+      });
+      return;
+    }
+
+    const finalPath = path.join(writeDir, '/fields.json');
+
+    return fieldsArrayToJson(fields).then(json => {
+      if (!fs.existsSync(writeDir)) {
+        fs.mkdirSync(writeDir, { recursive: true });
+      }
+      fs.writeFileSync(finalPath, json);
+
+      logger.success(
+        i18n(`${i18nKey}.converted`, {
+          src: dirName + `/${baseName}`,
+          dest: dirName + '/fields.json',
+        })
+      );
+      process.send({
+        action: 'COMPLETE',
+        finalPath,
+      });
+    });
+  });
+});
+
+/*
+ * Polyfill for `Array.flat(Infinity)` since the `flat` is only available for Node v11+
+ * https://stackoverflow.com/a/15030117
+ */
+function flattenArray(arr) {
+  return arr.reduce((flat, toFlatten) => {
+    return flat.concat(
+      Array.isArray(toFlatten) ? flattenArray(toFlatten) : toFlatten
+    );
+  }, []);
+}
+
+/**
+ * Converts an array of field objects to JSON
+ * @param {array} fields - Array of field objects
+ * @returns {string} - Returns converted JSON
+ */
+async function fieldsArrayToJson(fields) {
+  fields = await Promise.all(flattenArray(fields));
```

Comment:
Might be missing something, but why do we need `await Promise.all` here? It doesn't look like `flattenArray` returns a promise to me.

Reply:
Good Q. I Believe `fields` items can be promises so this just makes sure they all get resolved

Reply:
Ah, found this explanation: https://github.com/HubSpot/hubspot-cli/pull/738/files#r960887568